packages/devtools_app/lib/src/debugger/program_explorer.dart

```diff
@@ -0,0 +1,407 @@
+    Color color = colorScheme.functionSyntaxColor;
+    bool isShortCharacter;
+    if (object is ClassRef) {
+      character = 'c';
+      isShortCharacter = true;
+      color = colorScheme.declarationsSyntaxColor;
+    } else if (object is FuncRef) {
+      character = 'm';
+      isShortCharacter = true;
+      color = colorScheme.functionSyntaxColor;
+    } else if (object is FieldRef) {
+      character = 'f';
+      isShortCharacter = false;
+      color = colorScheme.variableSyntaxColor;
+    } else if (object is LibraryRef) {
+      icon = Icons.book;
+      color = colorScheme.modifierSyntaxColor;
+    } else if (object is ScriptRef) {
+      icon = libraryIcon;
+      color = colorScheme.stringSyntaxColor;
+    } else {
+      icon = containerIcon;
+    }
+
+    assert((icon == null && character != null && isShortCharacter != null) ||
+        (icon != null && character == null && isShortCharacter == null));
+
+    return SizedBox(
+      height: defaultIconSize,
+      width: defaultIconSize,
+      child: Container(
+        width: defaultIconSize,
+        height: defaultIconSize,
+        decoration: icon == null
+            ? BoxDecoration(
+                color: color,
+                shape: BoxShape.circle,
+              )
+            : null,
+        child: icon == null
+            ? Center(
+                child: Text(
+                  character,
+                  style: TextStyle(
+                    height: 1,
+                    fontFamily: theme.fixedFontStyle.fontFamily,
+                    color: theme.colorScheme.defaultBackgroundColor,
+                    fontSize: chartFontSizeSmall,
+                  ),
+                  // Required to center the individual character within the
+                  // shape. Since letters like 'm' are shorter than letters
+                  // like 'f', there's padding applied to the top of shorter
+                  // characters in order for everything to align properly.
+                  // Since we're only dealing with individual characters, we
+                  // want to disable this behavior so shorter characters don't
+                  // appear to be slightly below center.
+                  textHeightBehavior: TextHeightBehavior(
+                    applyHeightToFirstAscent: isShortCharacter,
+                    applyHeightToLastDescent: false,
+                  ),
+                ),
+              )
+            : Icon(
+                icon,
+                size: defaultIconSize,
+                color: color,
+              ),
+      ),
+    );
+  }
+}
+
+/// Picker that displays the program's structure, allowing for navigation and
+/// filtering.
+class ProgramExplorer extends StatelessWidget {
+  ProgramExplorer({
+    Key key,
+    @required this.debugController,
+    @required this.onSelected,
+  }) : controller = programExplorerController, super(key: key) {
+    controller.selectScriptNode(debugController.currentScriptRef.value);
```

Comment:
this should be handled in the controller, not in the widget constructor. The ProgramExplorer widget should not need to do this on initialization, but rather just read data from a value notifier in the controller.